src/gravel/controllers/nodes/deployment.py

```diff
@@ -124,3 +148,120 @@ def mark_ready(self) -> None:
         assert not self.error
         self._stage = NodeStageEnum.READY
         self._save_stage()
+
+
+class NodeDeployment:
+
+    _state: DeploymentState
+    _gstate: GlobalState
+    _connmgr: ConnMgr
+
+    def __init__(
+        self,
+        gstate: GlobalState,
+        connmgr: ConnMgr
+    ):
+        self._gstate = gstate
+        self._connmgr = connmgr
+        self._state = DeploymentState(gstate)
+
+    @property
+    def state(self) -> DeploymentState:
+        return self._state
+
+    async def join(
+        self,
+        leader_address: str,
+        token: str,
+        uuid: UUID,
+        hostname: str,
+        address: str
+    ) -> bool:
+        logger.debug(
+            f"join > with leader {leader_address}, token: {token}"
+        )
+
+        assert self._state
+        assert hostname
+        assert address
+
+        if self._state.bootstrapping:
+            raise NodeBootstrappingError()
+        elif self._state.deployed:
+            raise NodeHasBeenDeployedError()
+        elif self._state.joining:
+            raise NodeAlreadyJoiningError()
+        elif self._state.ready:
+            raise NodeHasJoinedError()
+        assert self._state.nostage
+
+        uri: str = f"ws://{leader_address}/api/nodes/ws"
+        conn = await self._connmgr.connect(uri)
+        logger.debug(f"join > conn: {conn}")
+
+        joinmsg = JoinMessageModel(
+            uuid=uuid,
+            hostname=hostname,
+            address=address,
+            token=token
+        )
+        msg = MessageModel(type=MessageTypeEnum.JOIN, data=joinmsg.dict())
+        await conn.send(msg)
+
+        self._state.mark_join()
+
+        reply: MessageModel = await conn.receive()
+        logger.debug(f"join > recv: {reply}")
+        if reply.type == MessageTypeEnum.ERROR:
+            errmsg = ErrorMessageModel.parse_obj(reply.data)
+            logger.error(f"join > error: {errmsg.what}")
+            await conn.close()
+            self._state.mark_error()
+            return False
+
+        assert reply.type == MessageTypeEnum.WELCOME
+        welcome = WelcomeMessageModel.parse_obj(reply.data)
+        assert welcome.pubkey
+        assert welcome.cephconf
+        assert welcome.keyring
+        assert welcome.etcd_peer
```

Comment:
wouldn't pydantic raise during `parse_obj` for these?

Reply:
it would, but given they're strings, it wouldn't catch an empty string.